prober/main.go

```diff
@@ -0,0 +1,266 @@
+
+	gotBQEntry, err := getFirstMatchedBQEntryWithRetries(ctx, bqQuery, cfg)
+	if err != nil {
+		return fmt.Errorf("failed to get match bigquery result: %w", err)
+	}
+
+	wantResourceMapping := &v1alpha1.ResourceMapping{
+		Resource: &v1alpha1.Resource{
+			Provider: "gcp",
+			Name:     fmt.Sprintf("//storage.googleapis.com/%s", cfg.GCSBucketID),
+		},
+		Contacts: &v1alpha1.Contacts{Email: []string{"prober@pmap.com"}},
+	}
+
+	gotPmapEvent := &v1alpha1.PmapEvent{}
+	if err := protojson.Unmarshal([]byte(gotBQEntry), gotPmapEvent); err != nil {
+		return fmt.Errorf("failed to unmarshal bigquery result to pmapevent: %w", err)
+	}
+
+	resourceMapping := &v1alpha1.ResourceMapping{}
+	if err := gotPmapEvent.GetPayload().UnmarshalTo(resourceMapping); err != nil {
+		return fmt.Errorf("failed to unmarshal to resource mapping: %w", err)
+	}
+
+	cmpOpts := []cmp.Option{
+		protocmp.Transform(),
+		protocmp.IgnoreFields(&v1alpha1.ResourceMapping{}, "annotations"),
+	}
+
+	var diffErr error
+	if diff := cmp.Diff(wantResourceMapping, resourceMapping, cmpOpts...); diff != "" {
+		diffErr = errors.Join(diffErr, fmt.Errorf("resourcemapping(ignore annotation) unexpected diff (-want,+got):\n%s", diff))
+	}
+
+	wantGithubSource := &v1alpha1.GitHubSource{
+		RepoName:           testGithubRepoValue,
+		Commit:             testGithubCommitValue,
+		Workflow:           testWorkflowValue,
+		WorkflowSha:        testWorkflowShaValue,
+		WorkflowRunId:      testWorkflowRunID,
+		WorkflowRunAttempt: 1,
+	}
+
+	if diff := cmp.Diff(wantGithubSource, gotPmapEvent.GetGithubSource(), cmpOpts...); diff != "" {
+		diffErr = errors.Join(diffErr, fmt.Errorf("githubSource unexpected diff (-want, +got):\n%s", diff))
+	}
+
+	return diffErr
+}
+
+// uploadFile uploads a object to the GCS bucket.
+func uploadFile(ctx context.Context, bucket, object string, data io.Reader) error {
+	o := gcsClient.Bucket(bucket).Object(object)
+	o = o.If(storage.Conditions{DoesNotExist: true})
+
+	// Upload an object with storage.Writer.
+	wc := o.NewWriter(ctx)
+	wc.Metadata = map[string]string{
+		server.MetadataKeyGitHubCommit:       testGithubCommitValue,
+		server.MetadataKeyGitHubRepo:         testGithubRepoValue,
+		server.MetadataKeyWorkflow:           testWorkflowValue,
+		server.MetadataKeyWorkflowSha:        testWorkflowShaValue,
+		server.MetadataKeyWorkflowRunAttempt: testWorkflowRunAttempt,
+		server.MetadataKeyWorkflowRunID:      testWorkflowRunID,
+	}
+
+	if _, err := io.Copy(wc, data); err != nil {
+		return fmt.Errorf("failed to copy bytes: %w", err)
+	}
+	if err := wc.Close(); err != nil {
+		return fmt.Errorf("failed to close writer: %w", err)
+	}
+
+	return nil
+}
+
+// getFirstMatchedBQEntryWithRetries query BigQuery table to find and return the matching entry.
+// If no result is found, query will be retried with the retry config.
+func getFirstMatchedBQEntryWithRetries(ctx context.Context, bqQuery *bigquery.Query, cfg *config) (string, error) {
```

Comment:
I think this is just duplication of e2e tests. We need to consider to refactor it to get rid of duplication

Reply:
I don't think making prober depends on package `integration` is a good idea. To do that, we need to export the functions, and plus, prober doesn't need stuffs like `testing.TB` but that's required by the integration test.

Reply:
I don't mean you make prober depends on package integration and just simple export `getFirstMatchedBQEntryWithRetries`. Further refactoring could be needed to avoid the duplication. My point is to get rid of duplication. There are other ways to achieve it instead of just exporting the `getFirstMatchedBQEntryWithRetries`. For example, you can extract GCS and BQ utility function